tango/common/remote_utils.py

```diff
@@ -0,0 +1,311 @@
+import atexit
+import datetime
+import json
+import logging
+import tempfile
+import time
+from abc import abstractmethod
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Any, Dict, List, Optional, Union
+
+from tango.common import PathOrStr
+from tango.common.exceptions import TangoError
+from tango.step import Step
+from tango.step_info import StepInfo
+
+from .registrable import Registrable
+
+logger = logging.getLogger(__name__)
+
+
+class RemoteConstants:
+    """
+    Common constants to be used as prefixes and filenames in remote workspaces.
+    """
+
+    RUN_DATASET_PREFIX = "tango-run-"
+    RUN_DATA_FNAME = "run.json"
+    STEP_DATASET_PREFIX = "tango-step-"
+    STEP_INFO_FNAME = "step_info.json"
+    STEP_RESULT_DIR = "result"
+    STEP_GRAPH_DATASET_PREFIX = "tango-step-graph-"
+    STEP_EXPERIMENT_PREFIX = "tango-step-"
+    STEP_GRAPH_FILENAME = "config.json"
+    GITHUB_TOKEN_SECRET_NAME: str = "TANGO_GITHUB_TOKEN"
+    RESULTS_DIR: str = "/tango/output"
+    INPUT_DIR: str = "/tango/input"
+
+    @classmethod
+    def step_dataset_name(cls, step: Union[str, StepInfo, Step]) -> str:
+        return f"{cls.STEP_DATASET_PREFIX}{step if isinstance(step, str) else step.unique_id}"
+
+    @classmethod
+    def step_lock_dataset_name(cls, step: Union[str, StepInfo, Step]) -> str:
+        return f"{cls.step_dataset_name(step)}-lock"
+
+    @classmethod
+    def run_dataset_name(cls, name: str) -> str:
+        return f"{cls.RUN_DATASET_PREFIX}{name}"
+
+
+@dataclass
+class RemoteDataset:
+    """
+    Abstraction for all objects in remote workspaces. Conceptually, this can be thought of as a folder.
+    """
+
+    name: str
+    """
+    Name of the dataset.
+    """
+    dataset_path: str
+    """
+    Remote location url for the dataset.
+    """
+    created: datetime.datetime
+    """
+    Time of creation.
+    """
+    committed: bool
+    """
+    If set to True, no further changes to the dataset are allowed.
+    """
+
+
+@dataclass
+class RemoteFileInfo:
+    """
+    Abstraction for file objects in remote workspaces.
+
+    Note: this is just mirroring beaker right now. We may not need this level of abstraction.
+    """
+
+    path: str
+    """
+    Remote location url for the file.
+    """
+    digest: str
+    """
+    Hash string representing the file.
+    """
+    updated: datetime.datetime
+    """
+    Last update time of the file.
+    """
+    size: int
+    """
+    Size of the file.
+    """
+
+
+class RemoteDatasetConflict(TangoError):
+    """
+    Error denoting that the remote dataset already exists.
+    """
+
+    pass
+
+
+class RemoteDatasetNotFound(TangoError):
+    """
+    Error denoting that the remote dataset does not exist.
+    """
+
+    pass
+
+
+class RemoteDatasetWriteError(TangoError):
+    """
+    Error denoting that there was an issue writing the dataset to its remote location.
+    """
+
+    pass
+
+
+class RemoteClient(Registrable):
```

Comment:
It seems to me like this is one level too many as far as abstractions go. There is a lot of commonality between `BeakerWorkspace` and `GCSWorkspace`, and you don't want to copy and paste that, so I see how you got here. Can you achieve the same thing by having some "protected" abstract class members in `RemoteWorkspace` that are overridden in `BeakerWorkspace` and `GCSWorkspace`, so we don't need to build up this parallel hierarchy of `*Client` classes?

Reply:
The clients are used in both step caches and workspaces. I still feel it's a useful abstraction to have, especially since we want to add stuff for search and filtering. We have control over beaker and its api, but not so much on gcloud/gcsfs. Having this layer helps with restricting changes in one place.